client/angular/services/columnAssignmentService.js

```diff
@@ -46,9 +46,48 @@
         }
 
         function backfillColumns(columnList, columnDataList) {
-            return columnList.map(function(column, idx) {
-                return column.slice(0, columnDataList[idx].slots);
+            var filledColumnList = columnDataList.map(function() {
+                return [];
+            });
+            var freeSlots = columnDataList.map(function(column) {
+                return column.slots;
+            });
+            var unassignedPinned = [];
+            var overflow = [];
+            //gredily "fill" each column
+            columnList.forEach(function truncate(column, idx) {
+                column.forEach(function(tweet) {
+                    if (tweet.weight <= freeSlots[idx]) {
+                        filledColumnList[idx].push(tweet);
+                        freeSlots[idx] -= tweet.weight;
+                    } else {
+                        if (tweet.pinned) {
+                            unassignedPinned.push(tweet);
+                        } else {
+                            overflow.push(tweet);
+                        }
+                    }
+                });
+                //if there's another column to be processed put any remaining pinned tweets in it
+                if (columnList.length > idx) {
+                    columnList[idx + 1] = unassignedPinned.concat(columnList[idx + 1]);
```

Comment:
Also this potentially inserts more than the maximum number of tweets into a column if the number of unassigned pinned tweets is very high